src/common/secret/src/vault_client.rs

```diff
@@ -0,0 +1,424 @@
+                auth_role_id,
+                auth_secret_id,
+            } => Some(secret::secret_hashicorp_vault_backend::Auth::ApproleAuth(
+                secret::VaultAppRoleAuth {
+                    role_id: auth_role_id.clone(),
+                    secret_id: auth_secret_id.clone(),
+                },
+            )),
+        };
+
+        secret::SecretHashicorpVaultBackend {
+            addr: self.addr.clone(),
+            path: self.path.clone(),
+            field: self.field.clone(),
+            auth,
+            tls_skip_verify: self.tls_skip_verify,
+        }
+    }
+}
+
+impl HashiCorpVaultClient {
+    pub fn new(config: HashiCorpVaultConfig) -> Result<Self> {
+        let mut client_builder = Client::builder();
+
+        if config.tls_skip_verify {
+            client_builder = client_builder.danger_accept_invalid_certs(true);
+        }
+
+        let client = client_builder
+            .build()
+            .context("Failed to create HTTP client")?;
+
+        Ok(Self { client, config })
+    }
+
+    pub async fn get_secret(&self) -> Result<Vec<u8>> {
+        // Try to get secret, with retry logic for token invalidation
+        let mut force_refresh_token = false;
+
+        // Retry loop for handling token invalidation
+        for retry_count in 0..2 {
+            // Get token (either directly or via app role)
+            let token = self.get_token_internal(force_refresh_token).await?;
+
+            // Fetch secret from Vault
+            let url = format!(
+                "{}/v1/{}",
+                self.config.addr.trim_end_matches('/'),
+                self.config.path
+            );
+            let response = self
+                .client
+                .get(&url)
+                .header("X-Vault-Token", &token)
+                .send()
+                .await
+                .context("Failed to send request to Vault")?;
+
+            // Handle authentication failures - token may have been rotated/revoked
+            if (response.status() == 401 || response.status() == 403)
+                && retry_count == 0
+                && matches!(self.config.auth, HashiCorpVaultAuth::AppRole { .. })
+            {
+                // this case means the token changed during cache, need to trigger a refresh
+                force_refresh_token = true;
+                continue;
+            }
+
+            if !response.status().is_success() {
+                return Err(anyhow::anyhow!(
+                    "Vault API returned error status: {} - {}",
+                    response.status(),
+                    response.text().await.unwrap_or_default()
+                ));
+            }
+
+            // Success case - process the response and break out of retry loop
+            return self.process_secret_response(response).await;
+        }
+
+        // Should never reach here due to early returns, but adding unreachable for safety
+        unreachable!("")
```

Comment:
Prefer to return an internal error instead. Especially since it will run on the meta-node. At least the caller can handle the error.